src/main/java/org/broadinstitute/consent/http/service/UserService.java

```diff
@@ -454,12 +454,135 @@ public void hasValidActiveERACredentials(User user) {
+    }
+
+    // case 4
+    if (hasInstitutionMatchingEmailDomain
+        && hasLibraryCard
+        && !hasAssignedInstitutionInDatabase
+        && !hasMatchingInstitutionInDatabase) {
+      updateInstitutionForUser(user, institutionFromEmail.getId());
+      return dropLibraryCardForUser(user);
+    }
+    // case 6
+    if (hasInstitutionMatchingEmailDomain
+        && !hasLibraryCard
+        && hasAssignedInstitutionInDatabase
+        && !hasMatchingInstitutionInDatabase) {
+      return updateInstitutionForUserAndReturnUpdatedUser(user, institutionFromEmail.getId());
+    }
+    // case 8
+    if (hasInstitutionMatchingEmailDomain
+        && !hasLibraryCard
+        && !hasAssignedInstitutionInDatabase
+        && !hasMatchingInstitutionInDatabase) {
+      return updateInstitutionForUserAndReturnUpdatedUser(user, null);
+    }
+    // case 10
+    if (!hasInstitutionMatchingEmailDomain
+        && hasLibraryCard
+        && hasAssignedInstitutionInDatabase
+        && !hasMatchingInstitutionInDatabase) {
+      return dropLCAndInstitutionForUser(user);
+    }
+    // case 12
+    if (!hasInstitutionMatchingEmailDomain
+        && hasLibraryCard
+        && !hasAssignedInstitutionInDatabase
+        && !hasMatchingInstitutionInDatabase) {
+      dropLibraryCardForUser(user);
+      return findUserByEmail(user.getEmail());
+    }
+    // case 14
+    if (!hasInstitutionMatchingEmailDomain
+        && !hasLibraryCard
+        && hasAssignedInstitutionInDatabase
+        && !hasMatchingInstitutionInDatabase) {
+      return updateInstitutionForUserAndReturnUpdatedUser(user, null);
+    }
+
+    // do nothing cases 1, 5, 15, 16
+    return user;
+
+    // not possible cases: 3, 7, 9, 11, 13
+  }
+
+  private void updateInstitutionForUser(User user, Integer institutionId) {
+    userDAO.updateInstitutionId(user.getUserId(), institutionId);
+  }
+
+  private User dropLCAndInstitutionForUser(User user) {
+    updateInstitutionForUser(user, null);
+    libraryCardDAO.deleteAllLibraryCardsByUser(user.getUserId());
+    return findUserByEmail(user.getEmail());
+  }
+
+  private User updateInstitutionForUserAndReturnUpdatedUser(User user, Integer institutionId) {
+    updateInstitutionForUser(user, institutionId);
+    return findUserByEmail(user.getEmail());
+  }
+
+  private User dropLibraryCardForUser(User user) {
+    libraryCardDAO.deleteAllLibraryCardsByUser(user.getUserId());
+    return findUserByEmail(user.getEmail());
+  }
+
+  @VisibleForTesting
+  protected boolean hasInstitutionMatchingEmailDomain(Institution institution) {
+    return institution != null;
+  }
+
+  @VisibleForTesting
+  protected boolean hasLibraryCard(User user) {
+    return !user.getLibraryCards().isEmpty();
+  }
+
+  @VisibleForTesting
+  protected boolean hasMatchingInstitutionInDatabase(
+      Institution institutionFromEmail, Institution institutionFromDatabase) {
+    if (institutionFromEmail == null || institutionFromDatabase == null) {
```

Comment:
If either the institution from the map lookup or the institution found in the database for the user are null, this check returns false.  This covers case 11.